src/Orleans.Core/Networking/ConnectionManager.cs

```diff
@@ -0,0 +1,425 @@
+            {
+                try
+                {
+                    entry = this.GetOrCreateEntry(address, ref acquiredConnectionLock);
+
+                    // Do not add a connection multiple times.
+                    if (entry.Connections.Contains(connection)) return entry;
+
+                    return this.connections[address] = entry.WithConnections(entry.Connections.Add(connection)).WithLastFailure(default);
+                }
+                finally
+                {
+                    if (acquiredConnectionLock) entry.ReleaseLock();
+                }
+            }
+        }
+
+        public void OnConnectionTerminated(SiloAddress address, Connection connection)
+        {
+            this.Remove(address, connection);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void ThrowIfRecentFailure(SiloAddress address, ConnectionEntry entry)
+        {
+            TimeSpan delta;
+            if (entry.LastFailure.HasValue
+                && (delta = DateTime.UtcNow.Subtract(entry.LastFailure.Value)) < CONNECTION_RETRY_DELAY)
+            {
+                throw new ConnectionFailedException($"Unable to connect to {address}, will retry after {delta.TotalMilliseconds}ms");
+            }
+        }
+
+        private ConnectionEntry GetOrCreateEntry(SiloAddress address, ref bool locked)
+        {
+            lock (this.collectionLock)
+            {
+                if (!this.connections.TryGetValue(address, out var entry))
+                {
+                    // Initialize the entry for this endpoint
+                    entry = ConnectionEntry.CreateNew();
+                    locked = entry.TryLock();
+
+                    this.connections[address] = entry;
+                }
+                else
+                {
+                    locked = entry.TryLock();
+                }
+
+                return entry;
+            }
+        }
+
+        private async ValueTask<Connection> ConnectAsync(SiloAddress address)
+        {
+            try
+            {
+                if (this.trace.IsEnabled(LogLevel.Information))
+                {
+                    this.trace.LogInformation(
+                        "Establishing connection to endpoint {EndPoint}",
+                        address);
+                }
+
+                var connection = await this.connectionFactory.ConnectAsync(address.Endpoint, this.cancellation.Token);
+
+                _ = Task.Run(async () =>
+                {
+                    Exception error = default;
+                    try
+                    {
+                        using (this.BeginConnectionScope(connection))
+                        {
+                            await connection.Run();
+                        }
+                    }
+                    catch (Exception exception)
+                    {
+                        error = exception;
+                    }
+
+                    this.Remove(address, connection);
```

Comment:
Shouldn't this happen in DisposeAsync of the returned ConnectionContext? Is it ok to have handed out a connection that is no longer being tracked internally?